### DISCUSSION TITLE
Several questions from noob

### DISCUSSION THREAD

**Discussion by ikaryss** [mozman/ezdxf#545] [Q&A] [answered]
Hello there! I'm new to ezdxf and I'm not sure that I understand all concepts from documentation correctly, So I have several questions:

1) **plotting layers separately**:
I have copied from docs the code that plots my dxf file   
```python
import sys
import matplotlib.pyplot as plt
from ezdxf import recover
from ezdxf.addons.drawing import RenderContext, Frontend
from ezdxf.addons.drawing.matplotlib import MatplotlibBackend

# Safe loading procedure (requires ezdxf v0.14):
try:
    doc, auditor = recover.readfile('mod_draw.dxf')
except IOError:
    print(f'Not a DXF file or a generic I/O error.')
    sys.exit(1)
except ezdxf.DXFStructureError:
    print(f'Invalid or corrupted DXF file.')
    sys.exit(2)

# The auditor.errors attribute stores severe errors,
# which may raise exceptions when rendering.
if not auditor.has_errors:
    fig = plt.figure()
    ax = fig.add_axes([0, 0, 1, 1])
    ctx = RenderContext(doc)
    out = MatplotlibBackend(ax)
    modelspace1 = doc.modelspace()
    Frontend(ctx, out).draw_layout(modelspace1, finalize=True)
```  
and it plots this 
![your](https://user-images.githubusercontent.com/33479999/134894115-048e7dc2-2d81-468b-99d8-4de842744d6e.png)

But I want to plot only purple lines that belong to "purple_layer" layer. So I decide to change the plotting line from this

` Frontend(ctx, out).draw_layout(modelspace1, finalize=True)`

to this

`Frontend(ctx, out).draw_layout(modelspace1.query('*[layer=="purple_layer"]'), finalize=True)`

But I get an error. Can you please tell me what I'm doing wrong?

2) **half view**

how to get a new modelspace with half view of the original dxf, given that the x-axis runs in the middle? Maybe there is some simple way to do this with ezdxf

3) **change line attributes**

how can I access and change the coordinates of specific lines? For example, I want to move the top of the purple triangle 20 mm to the left. To do this, I need to refer to the x-coordinate of two specific lines (assuming that half of the view is shown) and change its value. How do I do this? How do I know that I am changing the coordinate of the line I want?

**Comment by mbway** [comment, marked as answer]
1. you can change layer visibility with `RenderContext.set_layers_state(layers, state)`. Here 'state' means 'visible'
1a. you can use `Frontend.draw_entities` if you want to draw a collection of entities such as the result of a query. Look at the source code for `draw_layout` to see what other things you might want to set if you use `draw_entities`
2. you mean filter out entities in the bottom half? and do you mean outputting that half as an image or as another dxf file? If you just want a viewport onto the top half there is [this question](https://github.com/mozman/ezdxf/discussions/451) but if you want to actually filter the entities then you will have to loop over them and remove them
3. again, I'm not exactly sure what you want, but you could use the cad viewer to find the handle of the line you want to use, then access it directly, or loop over all the lines and pick out the line you want in code (eg if you want to pick the right-most line or something). Bear in mind that ezdxf is not a CAD application and depending on the setup of the CAD file moving that line may not move anything else with it because different entities are not necessarily connected. Depending on what entity type your line has (eg LINE or LWPOLYLINE etc) will depend on exactly how to move it. Look at the ezdxf docs for the entity you want to change

**Reply by ikaryss** [reply]
Thank you so much! But I have one more question. How exactly can I change the coordinates of already existing lines? For example, I select some line like this:  
``` python
purple_layer = modelspace.query('LINE[layer=="purple_layer"]')
line = purple_layer[0]
```
So how can I change its coordinates?  
I can retrieve the start / end vector, that give me `ezdxf.acc.vector.Vec3` type value, but I can't edit it
``` python
line.dxf.start
```

**Reply by mozman** [reply]
Points and vectors are intentionally immutable, you need to assign a new vector object:

```Python

# if needed the Vec3 object can be imported for ezdxf.math
from ezdxf.math import Vec3

purple_layers = modelspace.query('LINE[layer=="purple_layer"]i')
# The appended "i" makes the search case insensitive, which is the 
# behavior of AutoCAD for table entries (LAYER, STYLE, LINETYPE, ...).

line = purple_layers.first  # get the first entity or None, does not raise IndexError() for empty results
if line is not None:
    start = line.dxf.start  # returns a Vec3 object

    # reset start by assigning a tuple
    line.dxf.start = (10, start.y, start.z)

    # replace a single vector component, does the same as above:
    start = start.replace(x=10)  # returns a new Vec3(10, start.y, start.z)

    # translation (shift) is a vector addition:
    start += (10, 20, 30)

    # scaling is a scalar multiplication:
    start *= 2

    # reset the start coordinate of "line"
    line.dxf.start = start

```
